### PR TITLE
Deprecate AsdfExtension, AsdfExtensionList, BuiltinExtension

### PR DESCRIPTION
**Pull request by braingram**
~~Currently based off of https://github.com/asdf-format/asdf/pull/1409 and should not be merged before that PR.~~

### CODE REVIEW

**Comment by braingram** [comment]
Leaving this as draft until #1409 is merged

File: asdf/__init__.py [hunk 1]
```diff
@@ -27,7 +27,17 @@
...
+
+def __getattr__(name):
+    if name == "AsdfExtension":
+        # defer import to only issue deprecation warning when
+        # asdf.AsdfExtension is used
+        from asdf import extension
+
+        return extension.AsdfExtension
+    msg = f"module {__name__!r} has no attribute {name!r}"
+    raise AttributeError(msg)
```
Comment (WilliamJamieson) [inline asdf/__init__.py:43]: This is nice.

**Review by WilliamJamieson — APPROVED** [review]
LGTM

**Comment by WilliamJamieson** [comment]
How did this not show up in #1433? From the astropy downstream

> FAILED astropy/astropy/io/misc/asdf/tags/fits/tests/test_fits.py::test_backwards_compat - asdf.exceptions.AsdfDeprecationWarning: asdf.types.format_tag is deprecated...
FAILED astropy/astropy/io/misc/asdf/tags/table/tests/test_table.py::test_backwards_compat - asdf.exceptions.AsdfDeprecationWarning: asdf.types.format_tag is deprecated...

**Comment by WilliamJamieson** [comment]
astropy downstream will be fixed by astropy/astropy#14436

**Pull request merged**